plugins/filter_aws/aws.c

```diff
@@ -704,12 +698,17 @@ static int get_ec2_metadata(struct flb_filter_aws *ctx)
     if (ctx->tags_enabled && !ctx->tags_fetched) {
         ret = get_ec2_tags(ctx);
         if (ret < 0) {
-            flb_plg_error(ctx->ins, "Failed to get instance EC2 Tags");
-            return ret;
-        }
-        for (i = 0; i < ctx->tags_count; i++) {
-            if (ctx->tag_is_enabled[i] == FLB_TRUE) {
-                ctx->new_keys++;
+            flb_plg_error(ctx->ins, "Failed to get instance EC2 Tags, will not retry");
+            if (ret == -3) {
```

Comment:
My initial version of this PR is a 'proposal'. Minimal effort to handle 500s. In reality it is just to start the discussion about the desired solution, but most likely not the solution itself.

> From looking at the code myself quickly, it looks like this is only returned if the config is invalid?
> So that's the one case we don't retry?

`-3` means that configuration is invalid and Fluent Bit will exit with an error code if that happens.

Yes, that's the only case we don't retry -- we simply exit the application if that happens. For all other cases, if fetching tags will fail for any reason other than invalid configuration, we won't retry fetching tags (and all flushed logs won't have the tags injected).

---

> if a user enables tagging or fixes IMDS, Fluent Bit will start working without restart

EC2 option change for instance metadata server and tag values are only refreshed with the EC2 restart, so user would have to restart the instance anyway to have an effect (it doesn't apply for Nitro instances though). I don't have as much knowledge about the IMDS, so it might be one of the cases when it can be fixed without the restart.

---

> 1. Just retry forever on each flush and keep emitting error messages. This ensures that if a user enables tagging or fixes IMDS, Fluent Bit will start working without restart. The downside is that you get a spam of errors which might increase your log cost.

Retrying on each flush might be an overkill, especially since these errors may not change. I admit, it would be the simplest approach, so that's a plus.

In my opinion, we should allow the user to configure the maximum retry rate (with sane defaults provided). I wouldn't worry about the spam of errors as there is really no reasonable alternative. If user would be able to define the maximum retry rate, indirectly it allows to control the amount of error logs. In the end, at best someone will notice these errors and fix the underlying issue.

> 2. Retry some set number of time

It is an interesting option, but if an issue would be fixed after the defined max counter of retries, then our Filter wouldn't load the tags. I would rather expect the application to retry indefinitely, but maybe with lower rate if error is persistent. This way, it at least has the chance to fix itself.

> 3. Retry at some fixed interval which user can configure. So may be by default its 5 minutes or something

Yeah, I also like 'Option 3'. Let's call this 'Fixed Interval Retries'.
@PettitWesley I will try to implement this option as the next version of this PR.

---

I may have a few more ideas/suggestions.

### NIT: Abstraction Layer

First of all, I don't think we should treat EC2 Tags in a special way. EC2 Tags are only special at one angle, that EC2 must have the instance-metadata-options enabled in order for this functionality to work. In other words, we shouldn't retry 404s for EC2 Tags. All other errors should be treated equally in comparison to fetching other metadata infrmation pieces (like instance id). If we intend to implement a new retries mechanism for EC2 tags, then I would keep in mind that in the future it might be reused for other metadata information groups. That's just to say we should try to provide a decent abstraction layer for the retries, but we will see how it plays out in practice.

### Retries with Exponential Backoff

In terms of the retries and proposal for 'Fixed Interval Retries'. Personally I usually consider exponential backoff whenever I attempt to add retries to a component. I wonder if it would provide some value in this case as well. The main advantage of the exponential backoff is that Fluent Bit would fix itself faster in case the underlying issue is short-lived/temporary. It might start with retries at every flush and if it will fail repeatedly, it would adjust the retry duration to the longer periods -- possibly even longer than we would specify for the Fixed Interval Retries as we had sufficient number of retries to suspect that the issue is likely still occurring.

Example where exp backoff would be better:
1. [t+0s] flush, ec2 tags failed => logs without tags
2. [t+1s] flush, retry ec2 tags failed => logs without tags
3. [t+2s] flush, retry ec2 tags success => **logs with tags**
4. [t+3s] flush => **logs with tags**
5. [t+4s] flush => **logs with tags**
5. [t+5s] flush => **logs with tags**

Whereas in the 'Fixed Interval Retries', assuming a default retry duration set to 5s, the logs pushed at t+{2,3,4}s would have no tags.

What do you think? Do you see some value in introducing more complexity and exponential backoff with the potential advantage of more logs with injected tags?

### Allow to block pushing logs if EC2 Tags weren't fetched

For some use cases (for transparency, mine included), it might be useful to block pushing logs until Filter AWS fetches successfully all information groups. I mean it would have to fetch every enabled information in the configuration -- not only EC2 tags, but for example also instance id.

The idea is that, the more instances we have, the less useful are logs without tags. At the moment, if Fluent Bit flushes logs without tags, I am not really able to attribute them to particular components and debug the issues accordingly. It might be an edge case and specialised use case, but I would propose to expose an option in the configuration that would fail the Filter AWS initialization if fetching one of the groups failed.


---

Let me know what you think about my suggestions. As I said, I will try to implement the 'Fixed Interval Retries' as the next version of the PR. However, if you agree with some of my suggestions, then we might add more complexity to the PR.

(Also, sorry for late reply, I somehow missed the notifications from this PR too...)

Reply:
> EC2 option change for instance metadata server and tag values are only refreshed with the EC2 restart, so user would have to restart the instance anyway to have an effect (it doesn't apply for Nitro instances though). I don't have as much knowledge about the IMDS, so it might be one of the cases when it can be fixed without the restart.


Ah I see... I was testing on a nitro based instance before... I thought this was how it always works. 


I like both your "fixed interval retries" (simpler user config and experience) and exponential backoff retries ideas. Backoff has the downside that as your retry time grows, if you do fix the issue, it takes a while for it to take effect. Also, configuring backoff formula is probably more complicated for the users? What would be your proposed config options? For FLB backoff, we have scheduler.cap and scheduler.base to configure the max retry time and the backoff formula, we could do the same here with sane defaults. https://docs.fluentbit.io/manual/administration/scheduling-and-retries


> For some use cases (for transparency, mine included), it might be useful to block pushing logs until Filter AWS fetches successfully all information groups. I mean it would have to fetch every enabled information in the configuration -- not only EC2 tags, but for example also instance id. 

May be we handle this with a special zero value for the timeout time?

Reply:
> it might be useful to block pushing logs until Filter AWS fetches successfully all information groups

> May be we handle this with a special zero value for the timeout time?

OK, this is a good idea. Do you want me to include this behavior of retries=0 in this PR?